File: tests/unprotect-demo.c
#include <stdio.h>
#include <string.h>
#include <assert.h>

#include <oscore_native/message.h>
#include <oscore_native/test.h>
#include <oscore/protection.h>
#include <oscore/contextpair.h>
#include <oscore/context_impl/primitive.h>

int main()
{
    oscore_msgerr_native_t msgerr;

    struct oscore_context_primitive primitive = {
        .aeadalg = 24,
        .common_iv = "d\xf0\xbd" "1MK\xe0<'\x0c+\x1c",

        .recipient_id_len = 0,
        .recipient_key = "\xd5" "0\x1e\xb1\x8d\x06xI\x95\x08\x93\xba*\xc8\x91" "A|\x89\xae\t\xdfJ8U\xaa\x00\n\xc9\xff\xf3\x87Q",
    };

    // A message from plugtest example 1 with ChaCha algorithm
    oscore_msg_native_t msg = oscore_test_msg_create();
    msgerr = oscore_msg_native_append_option(
            msg,
            9,
            (uint8_t*)"\x09\x00",
            2);
    assert(!oscore_msgerr_native_is_error(msgerr));

    uint8_t *payload;
    size_t payload_len;
    oscore_msg_native_map_payload(msg, &payload, &payload_len);
    assert(payload_len >= 32);
    memcpy(payload, "\x5c\x94\xc1\x29\x80\xfd\x93\x68\x4f\x37\x1e\xb2\xf5\x25\xa2\x69\x3b\x47\x4d\x5e\x37\x16\x45\x67\x63\x74\xe6\x8d\x4c\x20\x4a\xdb", 32);
    msgerr = oscore_msg_native_trim_payload(msg, 32);
    assert(!oscore_msgerr_native_is_error(msgerr));

    enum oscore_unprotect_request_result oscerr;

    // Uninitialized values to be populated
    oscore_oscoreoption_t header;
    oscore_requestid_t request_id;
    oscore_msg_protected_t unprotected;

    bool found_oscoreoption = false;
    oscore_msg_native_optiter_t iter;
    oscore_msg_native_optiter_init(msg, &iter);
    while (!found_oscoreoption) {
        uint16_t number;
        const uint8_t *value;
        size_t value_length;
        bool next_exists = oscore_msg_native_optiter_next(msg, &iter, &number, &value, &value_length);
        if (!next_exists) {
            break;
        }
        if (number == 9) {
            found_oscoreoption = true;
            bool parsed = oscore_oscoreoption_parse(&header, value, value_length);
            assert(parsed);
        }
    }
    assert(found_oscoreoption);

    oscore_context_t *secctx = NULL;
    oscerr = oscore_unprotect_request(msg, &unprotected, header, secctx, &request_id);

    assert(oscerr == OSCORE_UNPROTECT_REQUEST_DUPLICATE);

    oscore_test_msg_destroy(msg);
    return 0;
}

File: tests/native-rust/src/unprotect_demo.rs
//! Run the "unprotect-demo" case, but with the test code in Rust.
//!
//! Unlike the unprotect demo, this derives the key right away (depending on having an HKDF
//! algorithm, which generally is available given we have a Rust backend).

use core::mem::MaybeUninit;

use coap_message::{MessageOption, MinimalWritableMessage, ReadableMessage};

use liboscore::raw;

pub fn run() -> Result<(), &'static str> {
    // From OSCORE plug test, security context A
    let immutables = liboscore::PrimitiveImmutables::derive(
        liboscore::HkdfAlg::from_number(5).unwrap(),
        b"\x9e\x7c\xa9\x22\x23\x78\x63\x40",
        b"\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10",
        None,
        liboscore::AeadAlg::from_number(24).unwrap(),
        b"\x01",
        b"",
    )
    .unwrap();

    let mut primitive = liboscore::PrimitiveContext::new_from_fresh_material(immutables);

    let mut msg = coap_message_implementations::heap::HeapMessage::new();
    let oscopt = b"\x09\x00";
    msg.add_option(9, oscopt).unwrap();
    msg.set_payload(b"\x5c\x94\xc1\x29\x80\xfd\x93\x68\x4f\x37\x1e\xb2\xf5\x25\xa2\x69\x3b\x47\x4d\x5e\x37\x16\x45\x67\x63\x74\xe6\x8d\x4c\x20\x4a\xdb").unwrap();

    liboscore_msgbackend::with_heapmessage_as_msg_native(msg, |msg| {
        unsafe {
            let header = liboscore::OscoreOption::parse(oscopt).unwrap();

            let mut unprotected = MaybeUninit::uninit();
            let mut request_id = MaybeUninit::uninit();
            let ret = raw::oscore_unprotect_request(
                msg,
                unprotected.as_mut_ptr(),
                &header.into_inner(),
                primitive.as_mut(),
                request_id.as_mut_ptr(),
            );
            assert!(ret == raw::oscore_unprotect_request_result_OSCORE_UNPROTECT_REQUEST_OK);
            let unprotected = unprotected.assume_init();

            let unprotected = liboscore::ProtectedMessage::new(unprotected);
            assert!(unprotected.code() == 1);

            let mut message_options = unprotected.options().fuse();
            let mut ref_options = [(11, "oscore"), (11, "hello"), (11, "1")]
                .into_iter()
                .fuse();
            for (msg_o, ref_o) in (&mut message_options).zip(&mut ref_options) {
                assert!(msg_o.number() == ref_o.0);
                assert!(std::str::from_utf8(msg_o.value()) == Ok(ref_o.1));
            }
            assert!(
                message_options.next().is_none(),
                "Message contained extra options"
            );
            assert!(
                ref_options.next().is_none(),
                "Message didn't contain the reference options"
            );
            assert!(unprotected.payload() == b"");
        };
    });

    // We've taken a *mut of it, let's make sure it lives to the end
    drop(primitive);

    Ok(())
}
